src/account/KeylessAccount.ts

```diff
@@ -0,0 +1,360 @@
+   */
+  readonly publicKey: KeylessPublicKey;
+
+  /**
+   * The EphemeralKeyPair used to generate sign.
+   */
+  readonly ephemeralKeyPair: EphemeralKeyPair;
+
+  /**
+   * The claim on the JWT to identify a user.  This is typically 'sub' or 'email'.
+   */
+  readonly uidKey: string;
+
+  /**
+   * The value of the uidKey claim on the JWT.  This intended to be a stable user identifier.
+   */
+  readonly uidVal: string;
+
+  /**
+   * The value of the 'aud' claim on the JWT, also known as client ID.  This is the identifier for the dApp's
+   * OIDC registration with the identity provider.
+   */
+  readonly aud: string;
+
+  /**
+   * A value contains 31 bytes of entropy that preserves privacy of the account. Typically fetched from a pepper provider.
+   */
+  readonly pepper: Uint8Array;
+
+  /**
+   * Account address associated with the account
+   */
+  readonly accountAddress: AccountAddress;
+
+  /**
+   * The zero knowledge signature (if ready) which contains the proof used to validate the EphemeralKeyPair.
+   */
+  proof: ZeroKnowledgeSig | undefined;
+
+  /**
+   * The proof of the EphemeralKeyPair or a promise that provides the proof.  This is used to allow for awaiting on
+   * fetching the proof.
+   */
+  readonly proofOrPromise: ZeroKnowledgeSig | Promise<ZeroKnowledgeSig>;
+
+  /**
+   * Signing scheme used to sign transactions
+   */
+  readonly signingScheme: SigningScheme;
+
+  /**
+   * The JWT token used to derive the account
+   */
+  readonly jwt: string;
+
+  /**
+   * An event emitter used to assist in handling asycronous proof fetching.
+   */
+  private readonly emitter: EventEmitter<ProofFetchEvents>;
+
+  constructor(args: {
+    address?: AccountAddress;
+    ephemeralKeyPair: EphemeralKeyPair;
+    iss: string;
+    uidKey: string;
+    uidVal: string;
+    aud: string;
+    pepper: HexInput;
+    proofOrFetcher: ZeroKnowledgeSig | Promise<ZeroKnowledgeSig>;
+    proofFetchCallback?: ProofFetchCallback;
+    jwt: string;
+  }) {
+    super();
+    const { address, ephemeralKeyPair, uidKey, uidVal, aud, pepper, proofOrFetcher, proofFetchCallback, jwt } = args;
+    this.ephemeralKeyPair = ephemeralKeyPair;
+    this.publicKey = KeylessPublicKey.create(args);
+    this.accountAddress = address ? AccountAddress.from(address) : this.publicKey.authKey().derivedAddress();
+    this.uidKey = uidKey;
+    this.uidVal = uidVal;
+    this.aud = aud;
+    this.jwt = jwt;
+    this.emitter = new EventEmitter<ProofFetchEvents>();
+    this.proofOrPromise = proofOrFetcher;
+    if (proofOrFetcher instanceof ZeroKnowledgeSig) {
+      this.proof = proofOrFetcher;
+    } else {
+      if (proofFetchCallback === undefined) {
+        throw new Error("Must provide callback for async proof fetch");
+      }
+      this.emitter.on("proofFetchFinish", async (status) => {
+        await proofFetchCallback(status);
+        this.emitter.removeAllListeners();
+      });
+      this.init(proofOrFetcher);
+    }
```

Comment:
I find this flow quite confusing tbh

The current flow:
- in case we don't have a zkp, we can pass a promise to the constructor instead.
- `KeylessAccount` will handle the asynchronous result and events, which I don't think it should
- once done fetching, the `proof` field will be updated without the user explicitly setting it, and if we want to react to it we need to listen to the event

I think we should aim for a more straightforward flow, here's a proposal:

The flow:
```typescript
const args = { ephemeralPublicKey, jwt, pepper };

const publicKey = new AnyPublicKey(KeylessPublicKey.create(args));
const address = await aptos.lookupOriginalAccountAddress({ authenticationKey: publicKey.authKey() });
// .. the app can now use the address and public key to fetch data, but can't sign yet

// Asynchronously get proof. This can also be fetched at a later time in case the user closes the page and aborts the call (we need this)
const proof = await aptosClient.getZKProof(args);

// We instantiate the actual account (note we also handle rotated auth keys here, by passing the looked-up address)
const account = new KeylessAccount({ ...args, address, proof });
```

Reply:
We can follow-up in another PR, but maybe let's not make an official release until we clean up the flow, to avoid having to make a breaking change right after

Reply:
Keeping as is, this is supported- see the test "creates the keyless account via the static constructor and submits a transaction"